scripts/commands/gotoname.lua

```diff
@@ -0,0 +1,97 @@
+-----------------------------------
+require("scripts/globals/status")
+
+cmdprops =
+{
+    permission = 1,
+    parameters = "si"
+}
+
+function error(player, msg)
+    player:PrintToPlayer(msg)
+    player:PrintToPlayer("!gotoname <mob or npc name> (index)")
+end
+
+-----------------------------------
+-- func: getValidEntities
+-- desc: given a list of entities, returns the ones that have valid coordinates
+-----------------------------------
+local getValidEntities = function(entities)
+    local validEntities = {}
+    for i, entity in pairs(entities) do
+        local pos = entity:getPos()
+        local invalidCoordinates = (pos.x == 0 and pos.y == 0 and pos.z == 0)
+
+        -- only add entity if it is at valid coordinates
+        if not invalidCoordinates then
+            table.insert(validEntities, entity)
+        end
+    end
+
+    return validEntities
+end
+
+-----------------------------------
+-- func: goToEntity
+-- desc: teleports the player to the given NPC
+-----------------------------------
+local goToEntity = function(player, entity)
+    -- determine whether we need zoneId parameter
+    local gotoZone = nil
+    if entity:getZoneID() ~= player:getZoneID() then
+        gotoZone = entity:getZoneID()
+    end
+
+    -- display message
+    player:PrintToPlayer(string.format("Going to %s %s (%i).", entity:getName(), entity:getZoneName(), entity:getID()))
+
+    -- half a second later, go.  this delay gives time for previous message to appear
+    player:timer(500, function(playerArg)
+        playerArg:setPos(entity:getXPos(), entity:getYPos(), entity:getZPos(), entity:getRotPos(), gotoZone)
+    end)
+end
+
+function onTrigger(player, pattern, index)
+    -- validate pattern
+    if pattern == nil or pattern == "" then
+        error(player, "You must enter an NPC name")
+        return
+    end
+
+    local zone = player:getZone()
+    local unfilteredEntities = zone:queryEntitiesByName(pattern)
+    entities = getValidEntities(unfilteredEntities)
+
+    if index ~= nil and index > 0 and index <= #entities then
+        goToEntity(player, entities[index])
+        return
+    end
+
+    if #entities > 20 then
+        player:PrintToPlayer("Too many results. Narrow your query or specify an index to go to.")
+        return
+    end
+
+    if #entities == 0 then
+        if #unfilteredEntities > 0 then
+            player:PrintToPlayer(string.format("%s not spawned in current zone", pattern))
+        else
+            player:PrintToPlayer(string.format("%s not found in current zone", pattern))
+        end
+
+        return
+    end
+
+    if #entities == 1 then
+        goToEntity(player, entities[1])
+        return
+    end
+
+    player:PrintToPlayer("Multiple entities found. Use !gotoname <mob or npc name> <index> to choose:")
+    for i, entity in pairs(entities) do
+        player:PrintToPlayer(string.format("[%d]: %s %s (%s)", i, entity:getName(), entity:getZoneName(), entity:getID()))
```

Comment:
Can you build a multiline string and print it once? Otherwise every line is split with 'SYSTEM MESSAGE' and the results take up twice as much space in the chat log as needed

Reply:
if its hitting max size (as indicated by truncation) the only option is to use multiple prints instead.

Reply:
> You can use other messaging types, like SYSTEM_MESSAGE_3: the yellow text
you see everywhere. Most other message types dont have the SYSTEM MESSAGE
header to deal with

ah, I didn't even consider that the header might count against the total size. I don't know if it does tho. and if we'd got twenty entities to return that case might still exceed max. Probably have to build one print per each in the loop, but we'll still want to use system 3 I think to avoid excess log spamage.

Reply:
Using SYSTEM_MESSAGE_3 now.